docs/source/development/eep-02-typing.md

```diff
@@ -0,0 +1,1474 @@
+will be marked as independent of the core proposal and summarized in
+[aligning names](aligning-names).
+
+## Motivation and resources
+
+- [Writing Python like it's Rust](https://kobzol.github.io/rust/python/2023/05/20/writing-python-like-its-rust.html).
+  A very good blogpost that summarizes the drawbacks of "stringly-typed" Python code and
+  shows how to incorporate typing philosophies from Rust into Python projects. Read this
+  if you don't have time to read the other resources.
+- [Robust Python](https://www.oreilly.com/library/view/robust-python/9781098100650/), an
+  excellent book that discusses how to design code around types and provides an
+  introduction to static type checkers in Python.
+- [jax enhancement proposal](https://jax.readthedocs.io/en/latest/jep/12049-type-annotations.html)
+  for adopting static typing. It has a very good discussion on benefits of static
+  typing.
+- [Subclassing in Python Redux](https://hynek.me/articles/python-subclassing-redux/)
+  explains which types of subclassing are considered harmful and was very helpful for
+  designing this proposal.
+
+For a summary of the new design philosophy, see [here](design-philosophy).
+
+## Changes for optimization
+
+The following changes apply to all functions that are directly related to optimization,
+i.e. `maximize`, `minimize`, `slice_plot`, `criterion_plot`, `params_plot`,
+`count_free_params`, `check_constraints` and `OptimizeResult`.
+
+### The objective function
+
+#### Current situation
+
+The objective or criterion function is the function being optimized. Currently, it is
+called `criterion` in estimagic.
+
+The `criterion` function maps a pytree of parameters into a criterion value.
+
+An important feature of estimagic is that the same criterion function can work for
+scalar, least-squares and likelihood optimizers. Moreover, a criterion function can
+return additional data that is stored in the log file (if logging is active). All of
+this is achieved by returning a dictionary instead of just a scalar float.
+
+The conventions for the return value of the criterion function are as follows:
+
+- For the simplest case, where only scalar optimizers are used, `criterion` returns a
+  float or a dictionary containing the entry "value" which is a float.
+- For likelihood functions, `criterion` returns a dictionary that contains at least the
+  entry `"contributions"` which can be any pytree that can be flattened into a numpy
+  array of floats.
+- For least-squares problems, `criterion` returns a dictionary that contains at least
+  the entry `"root_contributions"` which can be any pytree that can be flattened into a
+  numpy array of floats.
+- In any case the returned dictionary can have an arbitrary number of additional keys.
+  The corresponding information will be stored in the log database (if logging is used).
+
+A few simple examples of valid criterion functions are:
```

Comment:
Can you add a link to the docs / tutorials? IIRC, this is being used there, else may be explain a little? I found it especially confusing since you open above with *the same criterion function can work for scalar, least-squares and likelihood optimizers*, so I was trying to map into that until realizing it has nothing to do with each other.

Reply:
Our docs are actually not great about this. Need to change that soon. But I will explain it a bit better in the text.